File: binance4j-core/src/main/java/com/binance4j/core/serialization/CandleDeserializer.java
package com.binance4j.core.serialization;

import java.io.IOException;

import com.binance4j.core.dto.Candle;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.ObjectCodec;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.node.ArrayNode;

/** {@link Candle} deserializer */
public class CandleDeserializer extends JsonDeserializer<Candle> {

	@Override
	public Candle deserialize(JsonParser jp, DeserializationContext ctx) throws IOException {

		ObjectCodec oc = jp.getCodec();
		ArrayNode node = (ArrayNode) oc.readTree(jp);

		return new Candle(node.get(0).asLong(), node.get(1).asText(), node.get(2).asText(), node.get(3).asText(), node.get(4).asText(), node.get(5).asText(),
				node.get(6).asLong(), node.get(7).asText(), node.get(8).asLong(), node.get(9).asText(), node.get(10).asText());
	}
}

File: core/src/main/kotlin/com/binance4j/core/serialization/CandleDeserializer.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.core.serialization

import com.binance4j.core.dto.Candle
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.TreeNode
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.node.ArrayNode
import java.io.IOException

/** [Candle] deserializer  */
class CandleDeserializer : JsonDeserializer<Candle>() {
    @Throws(IOException::class)
    override fun deserialize(jp: JsonParser, ctx: DeserializationContext): Candle {
        val oc = jp.codec
        val node = oc.readTree<TreeNode>(jp) as ArrayNode
        return Candle(
            node[0].asLong(), node[1].asText(), node[2].asText(), node[3].asText(),
            node[4].asText(), node[5].asText(),
            node[6].asLong(), node[7].asText(), node[8].asLong(), node[9].asText(),
            node[10].asText()
        )
    }
}